edgedb-tokio/src/tutorial.md

```diff
@@ -0,0 +1,493 @@
+    .await?;
+```
+
+# Sample queries
+
+## When cardinality is guaranteed to be 1
+
+Using the `.query()` method works fine for any cardinality, but returns a `Vec` of results. This query with a cardinality of 1 returns a `Result<Vec<String>>` which becomes a `Vec<String>` after the error is handled:
+
+```rust
+let query = "select 'Just a string'";
+let query_res: Vec<String> = client.query(query, &()).await?;
+```
+
+But if you know that only a single result will be returned, using `.query_required_single()` or `.query_single()` will be more ergonomic:
+
+```rust
+let query = "select 'Just a string'";
+let query_res: String = client.query_required_single(query, &()).await?;
+let query_res_opt: Option<String> = client.query_single(query, &()).await?;
+```
+
+## Using the `Queryable` macro
+
+The easiest way to unpack an EdgeDB query result is the built-in `Queryable` macro from the `edgedb-derive` crate. This turns queries directly into Rust types without having to match on a `Value` (more on the `Value` enum in [its own section](#the-value-enum)), cast to JSON, etc.
+
+```rust
+#[derive(Debug, Deserialize, Queryable)]
+pub struct QueryableAccount {
+    pub username: String,
+    pub id: Uuid,
+}
+
+let query = "with account := 
+    (insert Account {
+      username := 'User1'
+    }),
+    select account {
+      username,
+      id
+    };";
+let as_queryable_account: QueryableAccount = client
+    .query_required_single(query, &())
+    .await?;
+```
+
+Note: Field order within the shape of the query matters when using the `Queryable` macro. In the example before, a query is done in the order `id, username` instead of `username, id` as defined in the struct:
+
+```rust
+let query = "with account := 
+    (insert Account {
+        username := 'User1'
+    }),
+    select account {
+      id,
+      username
+    };";
+let wrong_order: Result<QueryableAccount, _> = client
+    .query_required_single(query, &())
+    .await;
+assert!(
+    format!("{wrong_order:?}")
+    .contains(r#"WrongField { unexpected: "id", expected: "username" }"#);
+);
+```
+
+You can use [cargo expand](https://github.com/dtolnay/cargo-expand) with the nightly compiler to see the code generated by the Queryable macro, but the minimal example repo also contains a somewhat cleaned up version of the generated code [here](https://github.com/Dhghomon/edgedb_rust_client_examples/blob/master/src/lib.rs#L12).
+
+## Passing in arguments
+
+Technically arguments can be passed in through any correctly formatted string, but this can be fragile and quickly gets awkward. Note the double `{{` curly braces as well as `' '` around the `movie_title` string to keep the EdgeDB compiler from seeing two separate `movie` and `title` tokens:
+
+```rust
+let movie_title = "Nice movie";
+let release_year = 2023;
+let query = format!(
+    "with movie_title := '{movie_title}',
+year := {release_year},
+movie := (insert Movie {{
+    title := movie_title,
+    release_year := year
+}}),
```

Comment:
Please don't do this. People will just copy-paste this example and not read the text.

Also we don't do this not because curly braces and quotes are ugly, but to prevent injections (google SQL Injection) and partially for performance.

If you want to demo how ugly it is, put a plaintext example, like this:
```
cli.query("with movie_title := 'Who\'s Afraid of Virginia Woolf?' ...")
```

And please mention that it's not just "you _should_ pass arguments as tuple" but you **have to**, this is the only valid way.

It's also probably a good idea to mention that we're working on `QueryArgs` derive for named arguments.

Reply:
The ugliness part is just a didactic memory trick so definitely not needed considering how important it is to keep users using arguments the proper way. To help with that I added a small example with arguments closer to the top so that readers will have something to copy and paste in right away. Then removing the format string part made the lower section on arguments way too short so moved the content on arguments in general inside the REPL down to that section. Here are the changes: https://github.com/edgedb/edgedb-rust/commit/3a5b78af4fb210d196b66c98a392f257588f47b7